Cancellation and timeouts tutorial review (#4618)
Kotlin/kotlinx.coroutines: docs/topics/cancellation-and-timeouts.md

Reviewer: ```suggestion
```

Reviewer: I think we can delete the Duration import 🤔

[//]: # (title: Cancellation and timeouts)

Cancellation lets you request to stop a coroutine before it completes.
It stops work that's no longer needed, such as when a user closes a window or navigates away in a user interface while a coroutine is still running.

> You can use it to release resources early and to stop a coroutine from accessing objects past their disposal.

Reviewer: ```suggestion

You can use cancellation to release resources early and to stop a coroutine from accessing objects past their disposal.
```

Reviewer: just an extra line break to break it up and then add cancellation to the sentence for clarity

> You can also use cancellation to stop long-running coroutines, for example, sending heartbeats, running scheduled tasks, updating a state to reflect the newest reading (like the clock UI), etc.

Reviewer: ```suggestion
You can also use it to stop long-running coroutines that perform repeated work, for example:

* Sending heartbeats
* Running scheduled tasks
* Updating a state to reflect the newest reading, such as in a clock UI
```

Reviewer: Not sure if the repeated work part helps, but these all feel like it fits that criterion 🤔

Author: It's only an intro, I don't want to draw attention to this by using a list.

suspend fun main() {
    withContext(Dispatchers.Default) {
        // Used as a signal that the coroutine has started running
        val started = CompletableDeferred<Unit>()
        
        val job: Job = launch {
            println("The coroutine has started")

            // Completes the CompletableDeferred,
            // signaling that the coroutine has started running
            started.complete(Unit)
            try {

>                 // Suspends indefinitely

Reviewer: ```suggestion

```

>                 // This call will never return unless the coroutine is canceled

Reviewer: ```suggestion
                // Suspends until the coroutine is canceled
```

Reviewer: WDYT would this work instead?

General note: We should avoid future tense unless necessary. The present tense is usually clearer because it describes behavior directly. Future tense markers like "will" can sometimes make behavior sound less certain, and they often add extra unnecessary words.

Author: Agreed with the general idea, yet here "will never return" is used to explain what "suspends indefinitely" means.

Reviewer: Hmm 🤔 I don't think we need suspends indefintely though. "Suspends until the coroutine is canceled" describes what's happening. 

If we want to keep it for clarity though let's describe the behavior with present tense: 

> // This call never returns unless the coroutine is canceled

            println("This line will never be executed")
        }
      
        // Waits for the coroutine to start before canceling it
        jobStarted.await()

>         // Cancels the coroutine, so delay() throws a CancellationException

Author: The first example should be short and easy to comprehend. No need to repeat the same code twice. Added a commend on `async` below instead.

{kotlin-runnable="true" id="manual-cancellation-example"}

In this example, [`CompletableDeferred`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-completable-deferred/) is used as a signal that the coroutine has started running.

> The coroutine calls `complete()` when it starts executing, and `await()` only returns once that `CompletableDeferred` is completed. This way, cancellation happens only after the coroutine has started running. Without this check, the coroutine could be canceled before it runs the code inside its block. You don't need to have this check to cancel a coroutine, but it is included here to have a consistent example which always prints the explanation lines.

Reviewer: ```suggestion
The coroutine calls `complete()` when it starts executing, and `await()` only returns once that `CompletableDeferred` is completed. This way, cancellation happens only after the coroutine has started running. Without this check, the coroutine may be canceled before it runs the code inside its block. You don't need this check with `CompletableDeferred` to cancel a coroutine. It's included here to make the example consistent, so the coroutine always starts and prints its messages before it's canceled.
```

Author: That's misleading, you don't need this check to cancel a coroutine regardless of CompletableDerrefed.

> Similarly, a coroutine created by `async` can be canceled, `val deferred = async { ... }`, `deferred.cancel()`. The `async` builder returns a `Deferred` handle, which inherits from `Job`. Hence, the cancellation works in exactly the same way for `Deferred` as it does for `Job`.

Reviewer: ```suggestion
```

Reviewer: I think we can remove these lines, because we already covered this concept at the beginning:

> The [.async()](https://github.com/Kotlin/kotlinx.coroutines/blob/9746d58d1e8f267bb81592fbf2a2e094e665f570/docs/topics/coroutines-basics.md#coroutinescope-async) function returns a [Deferred](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-deferred/), which implements Job and supports the same cancellation behavior.

Author: Dmitry asked me to include these lines, since I removed the identical example with async, and I agree that it's nice to re-iterate this and show the syntax.

Reviewer: 🤔 good point. I think without the example we could show the syntax, but I'd suggest a very minimal example instead. 

Similarly, you can manually cancel a coroutine created by the `async()` builder function:

```kotlin
val deferred = async { /* ... */ }
deferred.cancel()
```

### Cancellation propagation

[Structured concurrency](coroutines-basics.md#coroutine-scope-and-structured-concurrency) ensures that canceling a coroutine also cancels all of its children.

> This prevents child coroutines from working after the parent has been requested to stop.

Reviewer: ```suggestion
This prevents child coroutines from continuing their work after their parent coroutine is canceled.
```

Reviewer: (suggestion)
I think requested to stop works, it's a nod back to the change at the beginning, but reading it this feels more natural - wdyt?

## Make coroutines react to cancellation {id="cancellation-is-cooperative"}

In Kotlin, coroutine cancellation is _cooperative_.

> This means that coroutines only react to cancellation when they cooperate by [suspending](#suspension-points-and-cancellation), [checking for cancellation explicitly](#check-for-cancellation-explicitly), or using the special case [`yield()`](#yield-often-in-non-suspending-code) function.

Author: Yep, reworded and added `runInterruptible`. I meant to summarise all subsections in this section.

## Make coroutines react to cancellation {id="cancellation-is-cooperative"}

In Kotlin, coroutine cancellation is _cooperative_.

> This means that coroutines only react to cancellation when they cooperate by [suspending](#suspension-points-and-cancellation) or [checking for cancellation explicitly](#check-for-cancellation-explicitly). We will also cover some typical usage patterns such as periodically calling the [`yield()`](#yield-often-in-non-suspending-code) suspending function in CPU-intensive code, and using [`runInterruptible()`](#interrupt-blocking-code-when-coroutines-are-canceled) to interrupt blocking code when coroutines are canceled.

Reviewer 2: [Minor, style] The sentences aren't in agreement syntactically: "Coroutines react to cancellation in specific cases. We will also describe other things." The transition can be improved: "Coroutines react to cancellation in specific cases. There are standard approaches for making coroutines cancellable: ..."

Reviewer 2: It's also important that there is no general way to deal with blocking code—`runInterruptible` only deals with *some* specific instances of blocking code, and only on the JVM.

In certain scenarios, you may use the [`isActive`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/is-active.html) property or the [`ensureActive()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/ensure-active.html) function to check if the coroutine was canceled. `isActive` is `false` when the coroutine is canceled and `ensureActive()` throws `CancellationException` if the coroutine is canceled.

However, you almost never need to check for cancellation explicitly. In most cases you can use `yield()`.

### `yield()` often in non-suspending code

In CPU-intensive computations and suspending code which is unlikely to suspend, call `yield()` periodically to allow cancellations to propagate.

The [`yield()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/yield.html) function suspends the coroutine, releasing the thread and giving other coroutines a chance to run on it. Suspending the coroutine lets it check for cancellation and throw `CancellationException` if it's canceled. Without suspending, coroutines on the same thread run sequentially.

> Thus `yield()` achieves two goals: frees up resources to allow others to run, and allows cancellations to propagate.

Reviewer 2: "Resources" isn't the best wording, as it often refers to locks or files, and we don't want `yield()` to release either of those. The paragraph just below is the more precise and useful statement to the same effect. Also, cancellation propagation between coroutines happens without `yield`, `yield` only allows *reacting* to cancellation in a specific coroutine to which cancellation was already propagated. The goal of this paragraph is unclear to me, it looks like it only reiterates what was already said, but in less precise and useful terms.

### Check for cancellation explicitly

If a coroutine doesn't suspend for a long time, it doesn't stop when it's canceled.

> In certain scenarios, you may use the [`isActive`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/is-active.html) property or the [`ensureActive()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/ensure-active.html) function to check if the coroutine was canceled. `isActive` is `false` when the coroutine is canceled and `ensureActive()` throws `CancellationException` if the coroutine is canceled.

Reviewer 2: [Minor, style] It's surprising to me that this subsection is above `yield()`, since we agreed on treating `yield()` as more important. As far as I can see, there is no dependency between the two sections, so we may order them from most to least significant freely.

However, you almost never need to check for cancellation explicitly. In most cases you can use `yield()`.

### `yield()` often in non-suspending code

In CPU-intensive computations and suspending code which is unlikely to suspend, call `yield()` periodically to let the current coroutine check for cancellation.

The [`yield()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/yield.html) function suspends the coroutine, releasing its current thread and giving other coroutines a chance to run on it. Suspending the coroutine lets it check for cancellation and throw `CancellationException` if it's canceled.

Thus `yield()` achieves two goals: lets other coroutines run, and allows checking for cancellations.

> Use `yield` to allow other coroutines to run on the same thread or thread pool before one of them finishes:

Reviewer 2: [Minor, style] Looks like redundancy in directly neighboring sentences? The `Use yield` sentence may be removed without losing anything, it feels like.

These blocking functions can be interrupted, which stops them prematurely.
However, when you call them from a coroutine, cancellation doesn't interrupt the thread.

To interrupt the thread when canceling a coroutine, wrap the blocking code into the [`runInterruptible()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-interruptible.html) function:

```kotlin
import kotlinx.coroutines.*

//sampleStart
suspend fun main() {
    withContext(Dispatchers.Default) {
        val started = CompletableDeferred<Unit>()

>         val job = launch {

Reviewer 2: > I don't think the intended audience going to have this problem, and for simplicity I'd like to stick with `job`.

It looks like we view our audience differently, then. I'd estimate that quite a large portion of the audience doesn't start new projects from scratch but gets onboarded into an existing project, tries working with `kotlinx.coroutines` there, gives up and goes to tutorials. This means that they are likely to have some chaotic exposure to unpredictable parts of our API.

> when I hide these bits of the examples to fight indentation on the playground, it will no longer make sense whose child it is. It's just a coroutine now.

Any `launch` invocation without an explicit receiver is the child of the caller, by convention. If you see `something; launch { somethingElse() }` anywhere, this is a child coroutine (or a stylistic mistake!).

>         val job = launch {

Author: Agreed on the audience. Semantically, it doesn't make sense to indicate that it is a child, if we don't refer to a particular parent or trying to stress that certain coroutines are siblings, etc. If it's always a child, it doesn't matter that it's a child.

>         val job = launch {

Author: All in all, you want to eradicate confusion of `val job` vs `Job`, I want simple names for readability.

>         val job = launch {

Reviewer 2: > Semantically, it doesn't make sense to indicate that it is a child, if we don't refer to a particular parent or trying to stress that certain coroutines are siblings, etc. If it's always a child, it doesn't matter that it's a child.

A job is not always a child:

```kotlin
coroutineScope {
    // The *current* job (may or may not be someone's child),
    // but in any case, it's not the current coroutine's child
    val job = currentCoroutineContext().job
}

coroutineScope {
    // The *child* job of the current coroutine
    val job = launch { }
}
```

Whether a job is the job of the currently running coroutine, its child, or someone else's job is important in practice, and I'd use `job`, `childJob`, and `otherJob` for these scenarios, respectively.

Here's an example:

```kotlin
// Defines a coroutine scope that uses the UI thread
class ScreenWithButtons(private val scope: CoroutineScope) {
    // Should only be called from the UI thread
    fun loadAndUpdateButtons(filename: String) {
        scope.launch {

>             // withContext will check for cancellation before entering the block

Reviewer: ```suggestion
            // withContext() checks for cancellation before entering the block
```

        scope.cancel()
    }

> }

Author: YES!

## Timeout

> Timeouts allow you to automatically cancel a coroutine after a specified duration. They are useful for stopping operations that take too long, to avoid serving outdated results, mitigate slow connection, or simply free up resources.

Author: "keep your application responsive" is too generic, added more specific reasons

Author: Yeah, it's all a bit too generic. I meant that if you detect a slow connection via a timeout, you can choose to walk away from it and use cache instead. Reworded.

I don't get the react to network failures example. Isn't catching an exception a reaction to network failures?

## Timeout

> A timeout allows you to automatically cancel a coroutine after a specified duration. It is useful for stopping operations that take too long. For example, if a request to download a picture from a server times-out, you can choose to retry, or fallback to the local cache.

Reviewer 2: [Minor, style] It's preferable to keep sentences on separate lines, to help git tooling determine what changed when. It's line-based by default in most cases.

Author: Will apply to the whole file after all approvals.

## Timeout

> A timeout allows you to automatically cancel a coroutine after a specified duration. It is useful for stopping operations that take too long. For example, if a request to download a picture from a server times-out, you can choose to retry, or fallback to the local cache.

Reviewer: ```suggestion
A timeout allows you to automatically cancel a coroutine after a specified duration. You can use it to stop operations that take too long. 
For example, if a request to download an image from a server times out, you can retry the request or fall back to the local cache.
```

suspend fun slowOperation(): Int {
    try {
        delay(300.milliseconds)

>         return 1

Author: Simplifying. Too many random numbers, alongside the timeout numbers, make it hard to keep track. Even better approach could have been to return a string, to avoid numbers altogether.

Reviewer 2: In my view, it's the exact opposite. `0`, `1`, or `2` can reasonably come up during a real-world program execution, so the eye is accustomed to them. `5` or `14` jump out at you. If you see `14` in the output, you can quickly find the corresponding `14` in the code and figure out where it came from.
